### PR TITLE
shrinking the deprecated method span

### PR DESCRIPTION
**Pull request by 0xPoe**
part of https://github.com/rust-lang/rust/issues/85403

r? @estebank

### CODE REVIEW

File: compiler/rustc_passes/src/stability.rs [hunk 1]
```diff
@@ -828,7 +828,15 @@ impl Visitor<'tcx> for Checker<'tcx> {
 
     fn visit_path(&mut self, path: &'tcx hir::Path<'tcx>, id: hir::HirId) {
         if let Some(def_id) = path.res.opt_def_id() {
-            self.tcx.check_stability(def_id, Some(id), path.span, None)
+            let method_span = if path.segments.len() >= 2 {
```
Comment (hi-rustin) [inline compiler/rustc_passes/src/stability.rs:831]: Maybe other logic is needed here? Only for method? But I'm not sure how to do that.

File: compiler/rustc_passes/src/stability.rs [hunk 2]
```diff
@@ -828,7 +828,15 @@ impl Visitor<'tcx> for Checker<'tcx> {
...
+                match path.segments.last() {
+                    Some(s) => Some(s.ident.span),
+                    None => None,
+                }
+            } else {
+                None
+            };
```
Comment (estebank) [inline compiler/rustc_passes/src/stability.rs:838]:
```suggestion
            let method_span = path.segments.last().map(|s| s.ident.span);
```

Comment (hi-rustin) [reply compiler/rustc_passes/src/stability.rs:838]: If we do not determine the length we will get the following error warning:
```rust
error: use of deprecated struct `raw::TraitObject`: use pointer metadata APIs instead https://github.com/rust-lang/rust/issues/81513
  --> library/core/src/raw.rs:85:10
   |
85 | #[derive(Copy, Clone)]
   |          ^^^^
   |
   = note: `-D deprecated` implied by `-D warnings`

error: use of deprecated struct `raw::TraitObject`: use pointer metadata APIs instead https://github.com/rust-lang/rust/issues/81513
  --> library/core/src/raw.rs:85:16
   |
85 | #[derive(Copy, Clone)]
   |                ^^^^^

```

**Comment by hi-rustin** [comment]
@estebank Could you please take a look? Thanks!

File: compiler/rustc_passes/src/stability.rs [hunk 3]
```diff
@@ -828,7 +828,12 @@ impl Visitor<'tcx> for Checker<'tcx> {
 
     fn visit_path(&mut self, path: &'tcx hir::Path<'tcx>, id: hir::HirId) {
         if let Some(def_id) = path.res.opt_def_id() {
-            self.tcx.check_stability(def_id, Some(id), path.span, None)
+            let method_span = if path.segments.len() >= 2 {
+                path.segments.last().map(|s| s.ident.span)
+            } else {
+                None
+            };
```
Comment (camsteffen) [inline compiler/rustc_passes/src/stability.rs:835]:
```suggestion
            let method_span = match path.segments {
                [.., _, last] => last.ident.span,
                _ => None,
            };
```

Comment (estebank) [reply compiler/rustc_passes/src/stability.rs:835]: r=me after applying the above

Comment (hi-rustin) [reply compiler/rustc_passes/src/stability.rs]: Addressed.

**Comment by richkadel** [comment]
FYI, this PR causes invalid warnings in some cases, causing build issues for some projects. Please take a look at #86000, which includes an MCVE to see what I'm referring to. Thanks!

**Comment by estebank** [comment]
@hi-rustin This got reverted and will need to be re-landed after addressing the invalid warning.

**Comment by hi-rustin** [comment]
> @hi-rustin This got reverted and will need to be re-landed after addressing the invalid warning.

Thanks, I will try to fix it.

**Comment by hi-rustin** [comment]
> @hi-rustin This got reverted and will need to be re-landed after addressing the invalid warning.

@estebank Do you have any thoughts on this issue? Maybe we should judge in the visit_path that only the method is processed? But I'm not sure how to determine if a segment is a method or not.

**Comment by estebank** [comment]
You might be able to "just" look at the `path_segment.res` to see if it is `Some(Res::Def(DefKind::AssocFn | DefKind::AssocTy | DefKind::AssocConst, _)`.

**Pull request merged**